snuba/datasets/entity_subscriptions/processors.py

```diff
@@ -0,0 +1,72 @@
+from abc import abstractmethod
+from typing import Any, Mapping, Optional, Sequence, Type, Union, cast
+
+from snuba.query.composite import CompositeQuery
+from snuba.query.conditions import (
+    BooleanFunctions,
+    ConditionFunctions,
+    binary_condition,
+    combine_and_conditions,
+)
+from snuba.query.data_source.simple import Entity
+from snuba.query.exceptions import InvalidQueryException
+from snuba.query.expressions import Column, Expression, Literal
+from snuba.query.logical import Query
+from snuba.utils.registered_class import RegisteredClass
+
+
+class EntitySubscriptionProcessor(metaclass=RegisteredClass):
+    @classmethod
+    def config_key(cls) -> str:
+        return cls.__name__
+
+    @classmethod
+    def get_from_name(cls, name: str) -> Type["EntitySubscriptionProcessor"]:
+        return cast(Type["EntitySubscriptionProcessor"], cls.class_from_name(name))
+
+    @abstractmethod
+    def to_dict(self, metadata: Mapping[str, Any]) -> Mapping[str, Any]:
+        raise NotImplementedError
+
+    @abstractmethod
+    def process(
+        self,
+        query: Union[CompositeQuery[Entity], Query],
+        metadata: Mapping[str, Any],
+        offset: Optional[int] = None,
+    ) -> None:
+        raise NotImplementedError
+
+
+class AddColumnCondition(EntitySubscriptionProcessor):
+    def __init__(self, extra_condition_data_key: str, extra_condition_column: str):
+        self.extra_condition_data_key = extra_condition_data_key
+        self.extra_condition_column = extra_condition_column
+
+    def to_dict(self, metadata: Mapping[str, Any]) -> Mapping[str, Any]:
+        if self.extra_condition_data_key not in metadata:
+            raise InvalidQueryException
+        return {self.extra_condition_data_key: metadata[self.extra_condition_data_key]}
+
+    def process(
+        self,
+        query: Union[CompositeQuery[Entity], Query],
+        metadata: Mapping[str, Any],
```

Comment:
I don't really get why metadata has to get passed everywhere now. Surely `offset` below would be a part of metadata?

Reply:
This makes sense; however, I'm not too familiar with why this was the case. Looking at how we currently do this, `tick_upper_offset` is a key in top-level key accessed in the [codec](https://github.com/getsentry/snuba/blob/master/snuba/subscriptions/codecs.py#L105). `SubscriptionData` on the other hand is obtained by `scheduled_subscription_dict["task"]["data"]` where `scheduled_subscription_dict` is the decoded raw payload as a dictionary. Moving this key (top-level --> inner-level) into the new metadata attribute we have defined will inherently change the shape of how subscriptions look like today.